pkg/cmd/util/flags/flags.go

```diff
@@ -33,3 +35,96 @@ func Resolve(args map[string][]string, fn func(*pflag.FlagSet)) []error {
 	fn(fs)
 	return Apply(args, fs)
 }
+
+// ComponentFlag represents a set of enabled components used in a command line.
+type ComponentFlag struct {
+	enabled  string
+	disabled string
+	isSet    func() bool
+
+	calculated sets.String
+
+	allowed  sets.String
+	mappings map[string][]string
+}
+
+// NewComponentFlag returns a flag that represents the allowed components and can be bound to command line flags.
+func NewComponentFlag(mappings map[string][]string, allowed ...string) *ComponentFlag {
+	set := sets.NewString(allowed...)
+	return &ComponentFlag{
+		allowed:  set,
+		mappings: mappings,
+		enabled:  strings.Join(set.List(), ","),
+		isSet:    func() bool { return false },
+	}
+}
+
+// Disable marks the provided components as disabled.
+func (f *ComponentFlag) Disable(components ...string) {
+	f.Calculated().Delete(components...)
+}
+
+// Enabled returns true if the component is enabled.
+func (f *ComponentFlag) Enabled(name string) bool {
+	return f.Calculated().Has(name)
+}
+
+// Calculated returns the effective enabled list.
+func (f *ComponentFlag) Calculated() sets.String {
+	if f.calculated == nil {
```

Comment:
Using flags before you parse flags is undefined behavior.  We don't lock it
elsewhere.

On Fri, Feb 19, 2016 at 11:43 AM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/util/flags/flags.go
> <https://github.com/openshift/origin/pull/6920#discussion_r53484634>:
>
> > +	}
> > +}
> > +
> > +// Disable marks the provided components as disabled.
> > +func (f *ComponentFlag) Disable(components ...string) {
> > +	f.Calculated().Delete(components...)
> > +}
> > +
> > +// Enabled returns true if the component is enabled.
> > +func (f *ComponentFlag) Enabled(name string) bool {
> > +	return f.Calculated().Has(name)
> > +}
> > +
> > +// Calculated returns the effective enabled list.
> > +func (f *ComponentFlag) Calculated() sets.String {
> > +	if f.calculated == nil {
>
> Disable calling into this and potentially locking things pre-parse is
> what concerns me
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6920/files#r53484634>.
>